plugins/sla_check.py

```diff
@@ -0,0 +1,179 @@
+
+  output = []
+  for dag in parent_dags:
+    dag_module = load_dag_module(dag.full_filepath)
+
+    if dag_module and hasattr(dag_module, 'DAG_SLA'):
+      sla = dag_module.DAG_SLA
+      name = dag_module.DAG_NAME
+
+      output.append((name,sla))
+  return output
+
+# Computes the datetime today that yesterday's DAG execution is due to
+# be completed by.
+def calculate_sla_due(sla_str):
+  sla_delta = parse_timedelta(sla_str)
+
+  return datetime.now().replace(hour=0, minute=0, second=0, microsecond=0) + sla_delta
+
+# Takes in a dag_id and an SLA value and computes if the DAG is due now
+# and if it is due now, checks the latest dag_run and reports out of compliance
+# if it does not yet exist for yesterday or if it is still running.
+def check_compliance(dag_name, sla):
+  sla_due = calculate_sla_due(sla)
+
+  if datetime.now() < sla_due:
+    return True
+
+  yesterday = date.today() - timedelta(days=1)
+  dt = yesterday.strftime("%Y-%m-%d")
+
+  last_run = settings.Session.query(DagRun).filter(
+    DagRun.dag_id == dag_name,
+    DagRun.execution_date == dt
+  ).first()
+
+  return last_run and last_run.state != State.RUNNING
+
+# This takes in a String formatted \d+hr\d+m\d+s and creates a parsed out
+# timedelta object that resembles the time from midnight the DAG needs to
+# finish by in order to be compliant.
+def parse_timedelta(time):
+  parts = regex.match(time)
+  if not parts:
+    return
+  parts = parts.groupdict()
+  time_params = {}
+  for (name, param) in parts.iteritems():
+    if param:
+      time_params[name] = int(param)
+  return timedelta(**time_params)
+
+# Wraps the above functions into a single call to find only the SLA's which
+# are in violation of their configured SLA.
+def find_violations():
+  slas = find_slas()
+  return filter(lambda s: not check_compliance(s[0], s[1]), slas)
+
+class SLACheckView(BaseView):
+  # Diagnostic endpoint which lists out the discovered SLA's and
+  # what time they are due at today.
+  @expose('/')
+  def index(self):
+    logging.info("SLA Check index() called")
+    slas = find_slas()
+    slas_with_due = [(x[0], calculate_sla_due(x[1])) for x in slas]
+    return self.render("sla_check/index.html", slas=slas_with_due)
+
+  # Wraps the violations for DataDog's format, it will return 200 if
+  # there are no violations and 500 and a comma-separated string of
+  # dag_id's that are currently in violation.
+  @expose('/healthcheck')
+  def health_check(self):
+    logging.info("SLA Health Check index() called")
+
+    violations = find_violations()
+    violation_names = [x[0] for x in violations]
+
+    if len(violation_names) > 0:
+      return ', '.join(violation_names), 500
+
+    else:
+      return '', 200
+
+  # Same data as the DataDog endpoint but wrapped in nicer JSON with
+  # additional metadata. This endpoint is mainly useful for potential
+  # future data collecting.
+  @expose('/health')
+  def health_status(self):
```

Comment:
could refactor the common code out of /healthcheck and /health.  Also, you might use one method for both and return different results based on the 'Accept' header (that's what it's for).  By default it returns text.   If you ask for json, you get the nicer response.